content/blogposts/2023/building-conc-better-structured-concurrency-for-go.md

```diff
@@ -0,0 +1,326 @@
+
+![Puppies parallel programming](https://storage.googleapis.com/sourcegraph-assets/blog/upload_ee1c4226c42ba701cde2324025599202.png)
+
+This goal of "make it hard to use incorrectly" comes with tradeoffs. Specifically, it means some (potentially useful) features are unlikely to be added to conc because they would break this goal. Shortly after the library went public I got a request to [add channels to conc](https://github.com/sourcegraph/conc/issues/56) and while I can see how some advanced user of conc could get value from this, I'd rather keep to the "hard to misuse" goal wherever possible.
+
+![GitHub issues conversation](https://storage.googleapis.com/sourcegraph-assets/blog/upload_e8c695bc11b8f2a62ceafaefa0d354dd.png)
+
+## Challenges I faced while building conc
+
+Conc is quite a small library, coming in at just under 2000 lines of Go code, but it took a lot of tweaking to get it to a point where I was happy to release it publicly.
+
+![SCC conc statistics](https://storage.googleapis.com/sourcegraph-assets/blog/upload_622f8a97c46fe4a1742ac48e45a35c2d.png)
+
+(Calculated using [scc](https://github.com/boyter/scc).)
+
+### challenge #1: Keeping the API natural and unsurprising
+To me, the API feels very natural now and most of the tweaks I did were to extend the goal of making it hard to use incorrectly. From my experience with other libraries, and looking at other concurrency implementations for inspiration, I found that nearly all of them suffer from at least one of the following problems.
+
+* They feel **unnatural** to use 
+* They contain many **foot guns** or they make it easy to do dangerous things without realizing
+* They make it **hard to understand** what's actually happening under the hood
+* They contain **surprising** implementation details that catch users off guard
+
+Often I found that I had to reduce the feature set of Conc in order to keep it simple and safe. I've made some slight tweaks after getting some feedback from users that further these goals. 
+
+For example, at Sourcegraph we generally use `GO_MAX_PROCS` to limit Goroutines to the number of core available. For us, that's often a sane default, but Go's scheduler can happily spawn 1000s of routines. For a majority of users, having this somewhat arbitrary limit was **surprising**, so by following the goal of "follow the principle of least surprise" it made sense to remove this.
+
+### challenge #2: Discovering edge cases while handling panics
+
+Panics aren't meant to be used as a way of exception handling, but while writing `conc` I needed to be able to catch and handle panics. One thing I learned about Go in spite of coding in it for years is that when you intercept a panic with a `nil` value, it will look like nothing actually went wrong.
```

Comment:
```suggestion
Panics aren't meant to be used as a way of exception handling, but I need `conc` to behave well even when a user gives it code that panics. One thing I learned about Go in spite of coding in it for years is that when you intercept a panic with a `nil` value, it will look like nothing actually went wrong.
```